lib/class-wp-webfonts.php

```diff
@@ -0,0 +1,290 @@
+<?php
+/**
+ * Webfonts API class.
+ *
+ * @package Gutenberg
+ */
+
+/**
+ * Class WP_Webfonts
+ */
+class WP_Webfonts {
+
+	/**
+	 * An array of registered webfonts.
+	 *
+	 * @static
+	 * @access private
+	 * @var array
+	 */
+	private static $webfonts = array();
+
+	/**
+	 * An array of registered providers.
+	 *
+	 * @static
+	 * @access private
+	 * @var array
+	 */
+	private static $providers = array();
+
+	/**
+	 * Stylesheet handle.
+	 *
+	 * @var string
+	 */
+	private $stylesheet_handle = '';
+
+	/**
+	 * Init.
+	 */
+	public function init() {
+
+		// Register default providers.
+		$this->register_provider( 'local', 'WP_Webfonts_Provider_Local' );
+
+		// Register callback to generate and enqueue styles.
+		if ( did_action( 'wp_enqueue_scripts' ) ) {
+			$this->stylesheet_handle = 'webfonts-footer';
+			$hook                    = 'wp_print_footer_scripts';
+		} else {
+			$this->stylesheet_handle = 'webfonts';
+			$hook                    = 'wp_enqueue_scripts';
+		}
+		add_action( $hook, array( $this, 'generate_and_enqueue_styles' ) );
+
+		// Enqueue webfonts in the block editor.
+		add_action( 'admin_init', array( $this, 'generate_and_enqueue_editor_styles' ) );
```

Comment:
**question (non-blocking):** Do we need to run this function on every admin page? I know it only renders styles when `wp-block-library` is available. Is there any issue with using the `enqueue_block_editor_assets` hook?

Reply:
`enqueue_block_editor_assets` doesn't work in this case... I remember debugging this a few months ago but I no longer remember exactly _why_ it happens. `admin_init` on the other hand works flawlessly, and provides an extra safety buffer and future-proofs the implementation.
Adding as inline styles on the `wp-block-library` stylesheet was a good way to ensure consistency between the frontend and the editor, and on the editor side we just let WP handle _when_ exactly the styles will be added, since they are attached to the block-library styles.

Reply:
Thanks for the context.